app/controllers/spree/admin/reports_controller.rb

```diff
@@ -247,8 +247,10 @@ def my_suppliers
       end
 
       def suppliers_of_products_distributed_by(distributors)
-        distributors.map { |d| Spree::Product.in_distributor(d).includes(:supplier).all }.
-          flatten.map(&:supplier).uniq
+        supplier_ids = Spree::Product.in_distributors(distributors).
+          select('spree_products.supplier_id')
+
+        Enterprise.where(id: supplier_ids)
```

Comment:
I'm assuming you use this as a trick to get unique enterprises. Why don't you ask that the DB exactly that in the query above? grouping by supplier id replaces the lack of `#distinct` in Rails 3.2.

Reply:
It's about passing an active record relation to the subquery, instead of querying distributors, then using #map on the result, then running queries for each one in the list.